### PR TITLE
Story/MON-12

### PR DESCRIPTION
**Pull request by aa-kashk**

### CODE REVIEW

File: monitor/src/main/java/edu/asu/diging/monitor/core/db/impl/AppDbConnection.java [hunk 1]
```diff
@@ -61,7 +62,11 @@ public void updateLastAppTest(String appId, String appTestId) {
 	 */
 	@Override
 	public void delete(IApp element) {
+		for (INotificationRecipient recipient: element.getRecipients()) {
+		    recipient.getApps().remove(element);
+		}
         em.remove(element);
+
```
Comment (jdamerow) [inline monitor/src/main/java/edu/asu/diging/monitor/core/db/impl/AppDbConnection.java]: please fix formatting

File: monitor/src/main/java/edu/asu/diging/monitor/core/service/impl/NotificationManager.java [hunk 2]
```diff
@@ -37,16 +43,21 @@
...
 			throw new EmailAlreadyRegisteredException();
 		}
 		INotificationRecipient recipient = new NotificationRecipient();
+		List<App> apps = new ArrayList<>();
 		recipient.setName(name);
 		recipient.setEmail(email);
+		recipient.setApps(apps);
+        for (String id: appIds) {
+            apps.add((App)appManager.getApp(id));
+        }
```
Comment (jdamerow) [inline monitor/src/main/java/edu/asu/diging/monitor/core/service/impl/NotificationManager.java:59]: indentation

File: monitor/src/main/java/edu/asu/diging/monitor/core/service/impl/NotificationManager.java [hunk 3]
```diff
@@ -66,12 +77,20 @@ public void deleteRecipient(String email) {
...
-	@Override
-	public void sendNotificationEmails(IApp app, AppStatus previousStatus) {
-		INotificationRecipient[] recipients = dbConnection.getAllRecipients();
-		for (INotificationRecipient recipient : Arrays.asList(recipients)) {
-			emailManager.sendAppStatusNotificationEmail(recipient.getEmail(), app, previousStatus.toString(),
-					app.getLastAppTest().getStatus().toString());
-		}
-	}
+	@Override //change here
+    public void sendNotificationEmails(IApp app, AppStatus previousStatus) {
```
Comment (jdamerow) [inline monitor/src/main/java/edu/asu/diging/monitor/core/service/impl/NotificationManager.java:85]: indentation, also, what does the comment mean?

File: monitor/src/main/java/edu/asu/diging/monitor/core/service/impl/NotificationManager.java [hunk 4]
```diff
@@ -66,12 +77,20 @@ public void deleteRecipient(String email) {
...
+	    List<INotificationRecipient> recipients;
+        if (app.getRecipients().isEmpty()) {
+            recipients = Arrays.asList(dbConnection.getAllRecipients());
+        }
+        else {
+            recipients = new ArrayList<>();
+            for (INotificationRecipient recipient : app.getRecipients())
+                recipients.add(recipient);
```
Comment (jdamerow) [inline monitor/src/main/java/edu/asu/diging/monitor/core/service/impl/NotificationManager.java]: - please follow our style guide
- also, why do you add all recipients  into a new list?

File: monitor/src/main/java/edu/asu/diging/monitor/web/admin/AddRecipientController.java [hunk 5]
```diff
@@ -21,12 +25,22 @@
...
 
-	@RequestMapping(value = "/admin/recipients/add", method = RequestMethod.GET)
-	public String show(Model model) {
-		model.addAttribute("recipientForm", new RecipientForm());
-		return "admin/recipients/add";
-	}
+    @RequestMapping(value = "/admin/recipients/add", method = RequestMethod.GET)
+    public String show(Model model) {
+        RecipientForm recipientForm = new RecipientForm();
+        recipientForm.setApps(appManager.getApps().stream().map(x -> {
```
Comment (jdamerow) [inline monitor/src/main/java/edu/asu/diging/monitor/web/admin/AddRecipientController.java]: `x` is a pretty bad variable name. even in a lambda function, it should be somewhat related to the object e.g. `a` for app

File: monitor/src/main/java/edu/asu/diging/monitor/core/service/impl/NotificationManager.java [hunk 6]
```diff
@@ -37,16 +43,21 @@
...
 			return false;
 		}
 		if (dbConnection.getById(email) != null) {
 			throw new EmailAlreadyRegisteredException();
 		}
 		INotificationRecipient recipient = new NotificationRecipient();
+		List<App> apps = new ArrayList<>();
 		recipient.setName(name);
 		recipient.setEmail(email);
+		recipient.setApps(apps);
```
Comment (jdamerow) [inline monitor/src/main/java/edu/asu/diging/monitor/core/service/impl/NotificationManager.java]: merge with line 54

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Jenkins, test this please.

**Comment by jdamerow** [comment]
The select all checkbox just looks like an app without a name, that needs to be changed (it should not be part of the list and have a label). Also, the apps should show who is getting notified. And the recipient list should show what apps are being watched.

**Comment by jdamerow** [comment]
- It should say "all" for recipients if an app notification is send to everybody.
- recipients should only be shown after login (not sure, if said otherwise before, if so, I'm sorry for the confusion)
- when adding or editing an app, we need to be able to select who gets notified. right now if I would add an app, I couldn't select any existing users to get notifications, only all.
- the select/deselect all checkbox needs to be visually different from the other checkboxes

File: monitor/src/main/java/edu/asu/diging/monitor/core/db/impl/AppDbConnection.java [hunk 7]
```diff
@@ -33,21 +34,34 @@ public IApp getById(String id) {
...
             throw new UnstorableObjectException("App does not have an id.");
         }
-        em.persist(app);
+        if (app.getRecipients() != null)
+            for (INotificationRecipient recipient : app.getRecipients()) {
+                recipient.getApps().add((App) app);
+                em.merge(recipient);
+            }
+        else
+            em.persist(app);
```
Comment (jdamerow) [inline monitor/src/main/java/edu/asu/diging/monitor/core/db/impl/AppDbConnection.java]: please follow our style guide

File: monitor/src/main/java/edu/asu/diging/monitor/core/db/impl/AppDbConnection.java [hunk 8]
```diff
@@ -33,21 +34,34 @@ public IApp getById(String id) {
...
-	}
+    @Override
+    public IApp update(IApp app) {
+        if (app.getRecipients() != null) {
+            for (INotificationRecipient recipient : app.getRecipients()) {
+                recipient.getApps().add((App) app);
+                em.merge(recipient);
+            }
+        } else
+            em.persist(app);
```
Comment (jdamerow) [inline monitor/src/main/java/edu/asu/diging/monitor/core/db/impl/AppDbConnection.java]: same here

**Comment by jdamerow** [comment]
Jenkins, test this please.

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
not sure if it has been introduced earlier or with this PR, but editing  of an app doesn't seem to work (editing the url)

**Comment by coder-aakash994** [comment]
are you talking about the healthUrl because I am able to update it locally while editing.

File: monitor/src/main/java/edu/asu/diging/monitor/core/db/impl/AppDbConnection.java [hunk 9]
```diff
@@ -33,21 +34,36 @@ public IApp getById(String id) {
...
         }
-        em.persist(app);
+        if (app.getRecipients() != null)
+            for (INotificationRecipient recipient : app.getRecipients()) {
+                recipient.getApps().add((App) app);
+                em.merge(recipient);
+            }
+		else {
+			em.persist(app);
+		}
```
Comment (jdamerow) [inline monitor/src/main/java/edu/asu/diging/monitor/core/db/impl/AppDbConnection.java:49]: indentation

File: monitor/src/main/java/edu/asu/diging/monitor/core/db/impl/AppDbConnection.java [hunk 10]
```diff
@@ -33,21 +34,36 @@ public IApp getById(String id) {
...
+    @Override
+    public IApp update(IApp app) {
+        if (app.getRecipients() != null && app.getRecipients().size() != 0) {
+            for (INotificationRecipient recipient : app.getRecipients()) {
+                recipient.getApps().add((App) app);
+                em.merge(recipient);
+            }
+		} else {
+			em.merge(app);
+		}
```
Comment (jdamerow) [inline monitor/src/main/java/edu/asu/diging/monitor/core/db/impl/AppDbConnection.java:63]: indentation

File: monitor/src/main/java/edu/asu/diging/monitor/web/admin/forms/AppForm.java [hunk 11]
```diff
@@ -14,6 +16,23 @@
...
+    }
+    public void setRecipients(List<RecipientForm> recipients) {
+        this.recipients = recipients;
+    }
+    public List<String> getRecipientIds() {
+        return recipientIds;
+    }
+    public void setRecipientIds(List<String> recipientIds) {
+        this.recipientIds = recipientIds;
+    }
```
Comment (jdamerow) [inline monitor/src/main/java/edu/asu/diging/monitor/web/admin/forms/AppForm.java:33]: formatting

**Comment by jdamerow** [comment]
There is some kind of endless recursion happening now. When polling, it gets all recipients for an app, then all apps for a recipient, with all apps, with all recipients...

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Pull request closed without merging**